### ISSUE TITLE
Add support for checkpoints

### ISSUE THREAD

**Issue by donkirkby** [cfe-lab/Kive#771]
Occasionally, we need to run long jobs that take days or weeks to complete. It's really frustrating if we have some problem with the server while a long job is running. Either we have to wait for the job to finish or we have to start over after restarting the server.

Maybe the simplest approach would be to upload all the outputs after a failure, and copy the original outputs when rerunning. Then a container could use one of its outputs as a checkpoint.

There's some [discussion] of checkpoint support in Singularity, and someone posted a proof of concept that uses [DMTCP].

[discussion]: https://github.com/sylabs/singularity/issues/468
[DMTCP]: https://github.com/dmtcp/dmtcp

**Comment by rhliang** [comment]
Unfamiliar with these technologies but would be awesome to get this implemented.  Do you know how these relate to BLCR?  I know Slurm supports checkpointing through that.

**Comment by donkirkby** [comment]
Looks like Slurm has [deprecated] BLCR in favour of CRIU, DMTCP, or SCR. The first two are mentioned in the discussion I linked to in the issue description, but I don't have experience with any of them.

[deprecated]: https://slurm.schedmd.com/checkpoint_blcr.html

**Comment by donkirkby** [comment]
Can you summarize your research here, @rhliang?

**Comment by rhliang** [comment]
Hm, I guess I didn't put it on Slack, but most of what I knew was summed up by this presentation, which I assume is from some core Slurm developers:

https://slurm.schedmd.com/SLUG16/ciemat-cr.pdf

And there seems to be a newer one too (which is still 2 years old now, but somehow I didn't find it before):

https://slurm.schedmd.com/SLUG17/ciemat-cr.pdf

DMTCP appears to be the one that they're hitching their wagon to.
